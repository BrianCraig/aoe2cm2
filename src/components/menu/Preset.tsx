import * as React from "react";
import {Link} from "react-router-dom";

class Preset extends React.Component<object, object> {
    public render() {
        return (
            <div>
                <p>Description of a single preset</p>
                <Link to='/'>Go to index</Link>
            </div>
        );
    }
}

export default Preset;