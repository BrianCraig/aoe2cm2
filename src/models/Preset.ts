import Civilisation from "./Civilisation";
import Turn from "./Turn";
import {CivilisationEncoder} from "../util/CivilisationEncoder";
import {Assert} from "../util/Assert";
import Player from "./Player";
import Action from "./Action";
import Exclusivity from "./Exclusivity";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly NEW: Preset = new Preset('Change me', Civilisation.ALL, []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);

    public static readonly ILLEGAL: Preset = new Preset('Illegal Preset', Civilisation.ALL, [
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL),
    ]);

    public static readonly SIMPLE_PARALLEL: Preset = new Preset('Simple Parallel', Civilisation.ALL, [
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL),
        new Turn(Player.HOST, Action.PICK, Exclusivity.GLOBAL, false, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.GLOBAL),
    ]);

    public static readonly REVEAL_TEST: Preset = new Preset('Reveal Test', Civilisation.ALL, [
        new Turn(Player.HOST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.HOST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.NONE, Action.REVEAL_PICKS, Exclusivity.GLOBAL),
        new Turn(Player.HOST, Action.SNIPE, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.SNIPE, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.HOST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.PICK, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.HOST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.GUEST, Action.BAN, Exclusivity.NONEXCLUSIVE, true),
        new Turn(Player.NONE, Action.REVEAL_PICKS, Exclusivity.GLOBAL),
        new Turn(Player.NONE, Action.REVEAL_BANS, Exclusivity.GLOBAL),
        new Turn(Player.NONE, Action.REVEAL_SNIPES, Exclusivity.GLOBAL),
    ]);

    public static readonly HIDDEN_1V1: Preset = new Preset('Hidden 1v1', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL
    ]);

    public static readonly HIDDEN_2V2: Preset = new Preset('Hidden 2v2', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly HIDDEN_3V3: Preset = new Preset('Hidden 3v3', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly HIDDEN_4V4: Preset = new Preset('Hidden 4v4', Civilisation.ALL, [
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
    ]);

    public static readonly M_CIVS: Preset = new Preset('M Civs', [Civilisation.MAGYARS, Civilisation.MALAY, Civilisation.MALIANS, Civilisation.MAYANS, Civilisation.MONGOLS], [
        Turn.HOST_PICK, Turn.GUEST_PICK
    ]);

    public readonly name: string;
    public readonly encodedCivilisations: string;
    public readonly turns: Turn[];

    constructor(name: string, civilisations: Civilisation[], turns: Turn[] = []) {
        this.name = name;
        this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(civilisations);
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations: string, turns: Turn[] } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isString(preset.encodedCivilisations);
        return new Preset(preset.name, CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations), Turn.fromPojoArray(preset.turns));
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get civilisations(): Civilisation[] {
        return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
    }
}

export default Preset;