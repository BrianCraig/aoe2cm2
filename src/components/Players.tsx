import * as React from 'react';
import {default as ModelPlayer} from '../models/Player';
import '../pure-min.css'
import '../style2.css'
import Player from "./Player";
import Preset from "../models/Preset";

interface IProps {
    nameHost: string;
    nameGuest: string;
    preset: Preset;
}

class Players extends React.Component<IProps, object> {
    public render() {
        return (
            <div className="pure-g players">
                <Player preset={this.props.preset} player={ModelPlayer.HOST} name={this.props.nameHost}/>
                <Player preset={this.props.preset} player={ModelPlayer.GUEST} name={this.props.nameGuest}/>

                <div className="hidden">
                    <span id="drafter_msg_host_captain">Host Captain</span>
                    <span id="drafter_msg_guest_captain">Guest Captain</span>
                </div>
            </div>
        );
    }
}

export default Players;
