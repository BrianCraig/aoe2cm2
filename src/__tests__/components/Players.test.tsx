import {shallow} from "enzyme";
import * as React from "react";
import Players from "../../components/Players";
import Preset from "../../models/Preset";


it('CivGrid renders correctly', () => {
    const component = shallow(<Players nameHost={'Sneaky Saladin'} nameGuest={'Beastly Barbarossa'}
                                       preset={Preset.SAMPLE}/>);
    expect(component).toMatchSnapshot();
});
