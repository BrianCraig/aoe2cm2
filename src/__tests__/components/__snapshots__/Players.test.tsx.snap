// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CivGrid renders correctly 1`] = `
<div
  className="pure-g players"
>
  <Connect(I18nextWithTranslation)
    name="Sneaky Saladin"
    player="HOST"
    preset={
      Preset {
        "civilisations": Array [
          Civilisation {
            "gameVersion": 2,
            "name": "Aztecs",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Berbers",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Britons",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Burmese",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Byzantines",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Celts",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Chinese",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Ethiopians",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Franks",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Goths",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Huns",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Incas",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Indians",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Italians",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Japanese",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Khmer",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Koreans",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Magyars",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Malay",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Malians",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Mayans",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Mongols",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Persians",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Portuguese",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Saracens",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Slavs",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Spanish",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Teutons",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Turks",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Vietnamese",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Vikings",
          },
        ],
        "name": "Default Preset",
        "turns": Array [
          Turn {
            "action": "GLOBAL_BAN",
            "player": "HOST",
          },
          Turn {
            "action": "GLOBAL_BAN",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_BAN",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_BAN",
            "player": "GUEST",
          },
          Turn {
            "action": "REVEAL_ALL",
            "player": "NONE",
          },
          Turn {
            "action": "GLOBAL_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "GLOBAL_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "EXCLUSIVE_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "EXCLUSIVE_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "REVEAL_ALL",
            "player": "NONE",
          },
          Turn {
            "action": "HIDDEN_SNIPE",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_SNIPE",
            "player": "GUEST",
          },
          Turn {
            "action": "REVEAL_ALL",
            "player": "NONE",
          },
        ],
      }
    }
  />
  <Connect(I18nextWithTranslation)
    name="Beastly Barbarossa"
    player="GUEST"
    preset={
      Preset {
        "civilisations": Array [
          Civilisation {
            "gameVersion": 2,
            "name": "Aztecs",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Berbers",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Britons",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Burmese",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Byzantines",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Celts",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Chinese",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Ethiopians",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Franks",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Goths",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Huns",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Incas",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Indians",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Italians",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Japanese",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Khmer",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Koreans",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Magyars",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Malay",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Malians",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Mayans",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Mongols",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Persians",
          },
          Civilisation {
            "gameVersion": 4,
            "name": "Portuguese",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Saracens",
          },
          Civilisation {
            "gameVersion": 3,
            "name": "Slavs",
          },
          Civilisation {
            "gameVersion": 2,
            "name": "Spanish",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Teutons",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Turks",
          },
          Civilisation {
            "gameVersion": 5,
            "name": "Vietnamese",
          },
          Civilisation {
            "gameVersion": 1,
            "name": "Vikings",
          },
        ],
        "name": "Default Preset",
        "turns": Array [
          Turn {
            "action": "GLOBAL_BAN",
            "player": "HOST",
          },
          Turn {
            "action": "GLOBAL_BAN",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_BAN",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_BAN",
            "player": "GUEST",
          },
          Turn {
            "action": "REVEAL_ALL",
            "player": "NONE",
          },
          Turn {
            "action": "GLOBAL_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "GLOBAL_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "EXCLUSIVE_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "EXCLUSIVE_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_PICK",
            "player": "GUEST",
          },
          Turn {
            "action": "REVEAL_ALL",
            "player": "NONE",
          },
          Turn {
            "action": "HIDDEN_SNIPE",
            "player": "HOST",
          },
          Turn {
            "action": "HIDDEN_SNIPE",
            "player": "GUEST",
          },
          Turn {
            "action": "REVEAL_ALL",
            "player": "NONE",
          },
        ],
      }
    }
  />
  <div
    className="hidden"
  >
    <span
      id="drafter_msg_host_captain"
    >
      Host Captain
    </span>
    <span
      id="drafter_msg_guest_captain"
    >
      Guest Captain
    </span>
  </div>
</div>
`;
